### PR TITLE
Throw errors with relevant info while parsing binary snapshot in odsp driver

### PR DESCRIPTION
**Pull request by jatgarg**
Refer: https://github.com/microsoft/FluidFramework/issues/4589
Throw errors with relevant info while parsing binary snapshot in odsp driver. This provides start, end index and type of the particular node in the parsed array so on error we can know what we were reading.

### CODE REVIEW

**Comment by vladsud** [comment]
I'm not sure start & end index in initial binary payload would be useful, as in most cases it would be hard to get access to that payload (no access to file, new snapshots posted, etc.).
I feel like more valuable would be the intention, i.e. why code expected something (type of node) at particular place, and what it got. And possibly - full path in the tree to problematic place, and what is actually at that place (i.e. type of node that does not match expectations). Later may container PII, but had relatively easy solution - we can mark this data as potentially user provided, and allow collection only if data comes from internal rings.

**Comment by jatgarg** [comment]
Ping

**Comment by jatgarg** [comment]
Also when you talk about the path in the tree, are you talking about the snapshot(blobs/trees path) or the tree structure that we create while parsing the binary snapshot?

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts [hunk 1]
```diff
@@ -54,9 +60,9 @@ function readAndValidateHeaderSection(node: NodeCore): ISnapshotHeader {
  */
 function readDictionarySection(node: NodeCore) {
     const dictionary = new Array<string>();
-    for (const name of node) {
-        assert(name instanceof BlobCore, 0x211 /* "Mapping should be of type BlobCore" */);
-        dictionary.push(name.toString());
+    for (const childNode of node) {
+        assertBlobCoreInstance(childNode, "Mapping should be of type BlobCore");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotParser.ts:64]: We can decide it later, but we should consider if we want each instance of assertBlobCoreInstance to have unique assert tag. I think the answer is yes, as strings add up a lot to bundle.
The way we solve it in other places is a project (package) to have a file that lists all the names of APIs that needs to be tagged, and tagger understands it and can do it not just for "assert" but for all other names.
Can you please open a bug to follow up here?

@anthony-murphy - FYI.

File: packages/drivers/odsp-driver/src/test/zipItDataRepresentationTests.spec.ts [hunk 2]
```diff
@@ -139,4 +144,46 @@ describe("Tree Representation tests", () => {
         node2.addString("fourth");
         validate(2 + 5 + 2 + 2 + 6 + 2 + 5 + 2 + 2 + 6);
     });
+
+    it("blob instance test", async () => {
+        const blobNode = new BlobShallowCopy(new ReadBuffer(new Uint8Array()), 0, 0);
+        assertBlobCoreInstance(blobNode);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/test/zipItDataRepresentationTests.spec.ts:150]: should string be required for assertBlobCoreInstance()? In general, I do not like it for asserts (as we should strive for code be readable / self-explanatory and require fewer comments), but while we do not have tags, it feels like string is required, no?

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts [hunk 3]
```diff
@@ -336,3 +336,64 @@ export class TreeBuilder extends NodeCore {
...
+    throwBufferParseException(node, "Number", message);
+}
+
+function throwBufferParseException(
+    node: NodeTypes,
+    expectedNodeType: NodeType,
+    message?: string,
+): never {
+    const error = createOdspNetworkError(
+        `BufferParsingException: ${message}`,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:376]: message == undefined would make it not the best string to read :)

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts [hunk 4]
```diff
@@ -336,3 +336,64 @@ export class TreeBuilder extends NodeCore {
...
+        fetchIncorrectResponse,
+        undefined,
+        undefined,
+        undefined,
+        {
+            nodeType: getNodeType(node),
+            expectedNodeType,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:383]: Is there any way to get a path in the tree logged? It's should be marked as potentially containing PII, but we can solve that problem separately - for DF / 1st party solutions that might be invaluable information to log.

That said, we can solve it separately, even having first level of data is better than nothing.

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 1 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts:64]: Created an issue  here: https://github.com/microsoft/FluidFramework/issues/7185

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts (hunk 4 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:383]: Created an issue here: https://github.com/microsoft/FluidFramework/issues/7186

**Pull request merged**